ESP-IDF support

bkbartk opened the pull request:
I have added ESP-IDF support,
with a little help of AI to be honest.

one of the important things is that for all other platforms, all functionality and code should remain exactly the same.
I have done this, tested this with lights and switches,
But I can only test the power(on/off) part, the rest I do not use, So I'm unable to test.

But for now I hope this can help people who are desperate en using ESP-IDF(like me)

Cossid commented:
I'm not sure when I'll get around to testing this, but here are my very rough initial thoughts:

First, ideally I'd like to see no changes to `device_groups.cpp` at all (the .h changes look fine).
The only differences I'm seeing are

- remapping away from sprintf_P/sprintf and PSTR to straight strings.  Please see if this was just an oversight and they are available, or if they can be remapped/redfined through headers instead
- `remoteIPAddress()` instead of `remoteIP()`, but since you're providing the full WiFiUdp implementation, including the header, you could simply just match those by changing the header/definition in WiFiUdp

I can't see the contents of CMakeLists.txt as it says it is binary, which is odd/concerning.

Lastly, since you are providing a full WiFiUdp class, I'm somewhat concerned this could conflict with anyone else using a similar base class on any other external component.  Perhaps it could be namespaced specifically for device_groups somehow.

Cossid commented:
Also, the `device_groups.h` change for `#include "esphome/components/network/ip_address.h"` looks unnecessary, and should just be removed from `wifiudp.h` instead perhaps.

Cossid commented:
That looks a lot closer to what I was expecting, with defines redirecting those functions.

I don't think we need CMakeLists.txt or component.mk at all, and it's probably best to remove them.
I think the new `esp_idf_compatibility.h` might benefit from a `#pragma once`
Then I think it may be in pretty good shape.

bkbartk commented:
yes, 
thank you,
I'm still working on it, the current one does not compile, because I missed `sprintf_P`
But I will fix this and let you know when everything is done,

bkbartk commented:
@Cossid 

I just made some changes, addressed all issues you mentioned, and it looks better,
I only tested on/off switches as that's the only thing I use, So I don't know how to test other stuff.
